lib/install/management/upgrade.go

```diff
@@ -0,0 +1,303 @@
+// limitations under the License.
+
+package management
+
+import (
+	"fmt"
+	"path"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/vmware/govmomi/object"
+	//	"github.com/vmware/govmomi/task"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/lib/config"
+	"github.com/vmware/vic/lib/install/data"
+	"github.com/vmware/vic/pkg/errors"
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/vsphere/extraconfig"
+	"github.com/vmware/vic/pkg/vsphere/extraconfig/vmomi"
+	"github.com/vmware/vic/pkg/vsphere/tasks"
+	"github.com/vmware/vic/pkg/vsphere/vm"
+
+	"golang.org/x/net/context"
+)
+
+const (
+	UpgradePrefix = "upgrade for"
+)
+
+// Upgrade will try to upgrade vch appliance to new version. If failed will try to roll back to original status.
+func (d *Dispatcher) Upgrade(vch *vm.VirtualMachine, conf *config.VirtualContainerHostConfigSpec, settings *data.InstallerData) (err error) {
+	defer trace.End(trace.Begin(conf.Name))
+
+	d.appliance = vch
+
+	// update the displayname to the actual folder name used
+	if d.vmPathName, err = d.appliance.FolderName(d.ctx); err != nil {
+		log.Errorf("Failed to get canonical name for appliance: %s", err)
+		return err
+	}
+
+	ds, err := d.session.Finder.Datastore(d.ctx, conf.ImageStores[0].Host)
+	if err != nil {
+		err = errors.Errorf("Failed to find image datastore %q", conf.ImageStores[0].Host)
+		return err
+	}
+	d.session.Datastore = ds
+
+	snapshotName := fmt.Sprintf("upgrade for %s", conf.Version.BuildNumber)
+	snapshotRefID, err := d.createSnapshot(snapshotName, "upgrade snapshot")
+	if err != nil {
+		return err
+	}
+
+	if err = d.update(conf, settings); err == nil {
+		return nil
+	}
+	log.Errorf("Failed to upgrade: %s", err)
+	log.Infof("Roll back upgrade")
+
+	// reset timeout, to make sure rollback still happens in case of deadline exceeded error in previous step
+	var cancel context.CancelFunc
+	d.ctx, cancel = context.WithTimeout(context.Background(), settings.RollbackTimeout)
+	defer cancel()
+
+	if rerr := d.rollback(conf, snapshotName); rerr != nil {
+		log.Errorf("Failed to revert appliance to snapshot: %s", rerr)
+		// return the error message for upgrade, instead of rollback
+		return err
+	}
+
+	// do clean up aggressively, even the previous operation failed with context deadline excceeded.
+	d.ctx = context.Background()
+	d.cleanSnapshot(*snapshotRefID, snapshotName)
+	d.deleteUpgradeImages(ds, settings)
+	log.Infof("Appliance is rollback to old version")
+	return err
+}
+
+func (d *Dispatcher) cleanSnapshot(id types.ManagedObjectReference, name string) error {
+	log.Infof("Deleting upgrade snapshot %q", name)
+	if _, err := tasks.WaitForResult(d.ctx, func(ctx context.Context) (tasks.Task, error) {
+		return d.appliance.RemoveSnapshot(ctx, id, true, true)
+	}); err != nil {
+		log.Errorf("Failed to clean up appliance upgrade snapshot %q: %s.", name, err)
+		log.Errorf("Be sure to remove snapshot %q manually before upgrade again", name)
```

Comment:
If it's not too hard maybe include the name of the VM so the user can navigate to the snapshot